File: 547. Friend Circles/solution.py
class Solution:
    def __init__(self):
        self.graph = {}
        self.visited = []
        
    def findCircleNum(self, M: List[List[int]]) -> int:
        # edge case
        if len(M) < 2:
            return len(M)
        
        # count friend circles
        count = 0
        self.visited = [False]*len(M)
        for i in range(len(M)):
            if not self.visited[i]:
                self.dfs(M, i)
                count += 1
                
        return count
        
    def dfs(self, M, i):
        if self.visited[i]:
            return 
        self.visited[i] = True
        for j in range(len(M[i])):
            if M[i][j] == 1:
                self.dfs(M, j)

                

File: 547. Friend Circles/solution.java
class Solution {
    public int findCircleNum(int[][] M) {
        // edge case
        int circles = 0;
        int n = M.length;
        if (n == 0) {
            return circles;
        }
        
        // visited list
        boolean[] visited = new boolean[n];
        for (int i = 0; i < n; i++) {
            visited[i] = false;
        }
        
        // dfs on each friend
        for (int i = 0; i < n; i++) {
            if (visited[i] == false) {
                circles++;
                dfs(i, M, visited);
            }
        }
        return circles;
    }
    
    public void dfs(int i, int[][] M, boolean[] visited) {
        if (visited[i] == true) {
            return;
        }
        
        visited[i] = true;
        for (int j = 0; j < visited.length; j++) {
            if (M[i][j] == 1)
                dfs(j, M, visited);
        }
    }
}
